### PR TITLE
pass rotary embedding to attention op

### PR DESCRIPTION
**Pull request by wangyems**

### CODE REVIEW

File: onnxruntime/contrib_ops/cpu/bert/attention_base.cc [hunk 1]
```diff
@@ -253,6 +253,7 @@ Status AttentionBase::CheckInputs(const TensorShape& input_shape,
     output_parameters->is_unidirectional = is_unidirectional_;
     output_parameters->past_present_share_buffer = (past_present_share_buffer_ != 0 && past != nullptr);
     output_parameters->do_rotary = do_rotary_;
+    output_parameters->rotary_embedding = rotary_embedding_ == 0 ? (int)(output_parameters->head_size) : rotary_embedding_;
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cpu/bert/attention_base.cc:256]: the default is head_size/2?

**Review by yufenglee — APPROVED** [review]

File: onnxruntime/contrib_ops/cuda/bert/add_bias_transpose.cu [hunk 2]
```diff
@@ -650,20 +650,20 @@ void InvokeAddBiasTranspose(
     if (format != 1 && format != 2 && format != 3) {
       ORT_THROW("format must be 1, 2 or 3 for rotary attention");
     }
-    if (qk_head_size != 64 && qk_head_size != 128) {
-      ORT_THROW("qk_head_size must be 64 or 128 for rotary attention");
+    if (rotary_embedding != 32 && rotary_embedding != 64 && rotary_embedding != 128) {
+      ORT_THROW("rotary_embedding must be 64 or 128 for rotary attention");
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/add_bias_transpose.cu:654]: update message for 32

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 3]
```diff
@@ -333,6 +333,10 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
               "Whether to use rotary position embedding. Default value is 0.",
               AttributeProto::INT,
               OPTIONAL_VALUE)
+        .Attr("rotary_embedding",
+              "Dimention of rotary embedding. Limited to 32 or 64. Default value is head_size",
```
Comment (tianleiwu) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc:337]: typo: Dimension. Need we add 128 here?

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 4]
```diff
@@ -333,6 +333,10 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
               "Whether to use rotary position embedding. Default value is 0.",
               AttributeProto::INT,
               OPTIONAL_VALUE)
+        .Attr("rotary_embedding",
```
Comment (tianleiwu) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc:336]: rename it to rotary_embedding_dim?

**Review by yufenglee — APPROVED** [review]
:shipit:

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 5]
```diff
@@ -333,6 +333,10 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
               "Whether to use rotary position embedding. Default value is 0.",
               AttributeProto::INT,
               OPTIONAL_VALUE)
+        .Attr("rotary_embedding_dim",
+              "Dimension of rotary embedding. Limited to 32, 64 or 128. Default value is head_size",
```
Comment (tianleiwu) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc:337]: where is the code for handling "Default value is head_size"?

File: onnxruntime/contrib_ops/cuda/bert/add_bias_transpose.cu [hunk 6]
```diff
@@ -650,20 +650,20 @@ void InvokeAddBiasTranspose(
     if (format != 1 && format != 2 && format != 3) {
       ORT_THROW("format must be 1, 2 or 3 for rotary attention");
     }
-    if (qk_head_size != 64 && qk_head_size != 128) {
-      ORT_THROW("qk_head_size must be 64 or 128 for rotary attention");
+    if (rotary_embedding != 32 && rotary_embedding != 64 && rotary_embedding != 128) {
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/add_bias_transpose.cu:653]: Can this be moved to check_inputs of the operator?

**Pull request merged**